test-harness/tests/test_cache.py

```diff
@@ -668,3 +668,249 @@ def test_if_early_boot_stage_can_recover_from_a_bit_slow_backend(
             with GuardedSubprocess(v):
                 time.sleep(backend_request_timeout * 0.3 + 1)  # let it warm-up!
                 ping_mesos_agent(ar, superuser_user_header)
+
+
+class TestCacheMarathon:
+
+    def test_upstream_wrong_json(
+            self, nginx_class, mocker, superuser_user_header):
+        filter_regexp = {
+            "Cannot decode Marathon apps JSON: ": SearchCriteria(1, True),
+        }
+
+        ar = nginx_class()
+
+        # Send wrong non-json response content
+        mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                            func_name='apps_response_raw',
+                            aux_data=b"wrong response")
+
+        url = ar.make_url_from_path('/service/nginx-enabled/foo/bar/')
+        with GuardedSubprocess(ar):
+            # Register Line buffer filter:
+            lbf = LineBufferFilter(filter_regexp,
+                                   timeout=5,  # Just to give LBF enough time
+                                   line_buffer=ar.stderr_line_buffer)
+
+            # Trigger cache update by issuing request:
+            resp = requests.get(url,
+                                allow_redirects=False,
+                                headers=superuser_user_header)
+            assert "cache state is invalid" in resp.content.decode('utf-8')
+            assert resp.status_code == 503
+
+            lbf.scan_log_buffer()
+
+        assert lbf.all_found
+        assert lbf.extra_matches == {}
+
+    def test_app_without_labels(
+            self, nginx_class, mocker, superuser_user_header):
+        filter_regexp = {
+            "Labels not found in app '/without_labels'": SearchCriteria(1, True),
+        }
+        apps = [{"id": "/without_labels"}]
```

Comment:
Here and in all other tests you are assuming certain order in which apps data is processed (`if`s to be precise) and this is `a very bad thing` ™️ .

Any refactoring that does not change the behaviour, just order in which data is processed, will make all your tests hard-fail with a confusing message, even though the functionality they are testing has not changed. What is more, we may get false negatives if something is breaking the code along the processing path, but the fact that apps data is incomplete makes the cache query fail anyway.

My strong suggestion here would be to re-use `NGINX_TASK_ALWAYSTHERE` data, which we can import and modify at will. The modifications that we are going to do should be as narrow as possible - break only one thing in the apps dict, and test the resulting behaviour.